doxygen_doc/man/man3/mjl_Edge.3
.TH "mjl::Edge" 3 "Thu May 29 2025 14:52:20" "Version 1.0" "Open-CSMP++" \" -*- nroff -*-
.ad l
.nh
.SH NAME
mjl::Edge
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBEdge\fP (const \fBmjl::Point\fP &, const \fBmjl::Point\fP &)"
.br
.ti -1c
.RI "\fBEdge\fP (const \fBmjl::Edge\fP &)"
.br
.ti -1c
.RI "\fBEdge\fP & \fBoperator=\fP (const \fBmjl::Edge\fP &m)"
.br
.ti -1c
.RI "void \fBSet\fP (const \fBmjl::Point\fP &, const \fBmjl::Point\fP &)"
.br
.ti -1c
.RI "void \fBSet\fP (double x1, double y1, double x2, double y2)"
.br
.ti -1c
.RI "\fBEdge\fP & \fBRot\fP ()"
.br
.RI "pivots an edge 90o clockwise around it's midpoint "
.ti -1c
.RI "\fBEdge\fP & \fBFlip\fP ()"
.br
.RI "flip the direction of an edge "
.ti -1c
.RI "void \fBMove\fP (double dx, double dy)"
.br
.ti -1c
.RI "void \fBScale\fP (double xfac, double yfac)"
.br
.ti -1c
.RI "void \fBNormalizeTo\fP (double n)"
.br
.ti -1c
.RI "bool \fBIfCloserThanMerge\fP (double tolerance, \fBmjl::Point\fP &)"
.br
.ti -1c
.RI "bool \fBoperator==\fP (const \fBmjl::Edge\fP &) const"
.br
.ti -1c
.RI "bool \fBoperator!=\fP (const \fBmjl::Edge\fP &) const"
.br
.ti -1c
.RI "bool \fBoperator<\fP (const \fBmjl::Edge\fP &) const"
.br
.ti -1c
.RI "\fBmjl::Point\fP \fBPoint\fP (double t) const"
.br
.ti -1c
.RI "double \fBY\fP (double x) const"
.br
.ti -1c
.RI "\fBmjl::Point\fP \fBOrigin\fP () const"
.br
.ti -1c
.RI "\fBmjl::Point\fP \fBDestination\fP () const"
.br
.ti -1c
.RI "void \fBMidPoint\fP (\fBmjl::Point\fP &) const"
.br
.ti -1c
.RI "\fBmjl::Point\fP \fBMidPoint\fP () const"
.br
.ti -1c
.RI "bool \fBIsVertical\fP () const"
.br
.ti -1c
.RI "double \fBSlope\fP () const"
.br
.ti -1c
.RI "double \fBDistanceFrom\fP (const \fBmjl::Point\fP &) const"
.br
.ti -1c
.RI "double \fBLength\fP () const"
.br
.ti -1c
.RI "LOCATION \fBClassify\fP (const \fBmjl::Point\fP &) const"
.br
.ti -1c
.RI "INTERSECTION \fBIntersect\fP (const \fBEdge\fP &, double &) const"
.br
.ti -1c
.RI "INTERSECTION \fBCross\fP (const \fBEdge\fP &, double &) const"
.br
.ti -1c
.RI "INTERSECTION \fBCrossingPoint\fP (double tolerance, const \fBEdge\fP &, \fBmjl::Point\fP &) const"
.br
.ti -1c
.RI "INTERSECTION \fBSKM_CrossingPoint\fP (double tolerance, const \fBEdge\fP &, \fBmjl::Point\fP &) const"
.br
.ti -1c
.RI "EDGE_CLASSIFICATION \fBType\fP (const \fBmjl::Point\fP &) const"
.br
.ti -1c
.RI "double \fBAngleTo\fP (const \fBEdge\fP &) const"
.br
.RI "see "Mathematische Formeln" p\&. 170, Schnittwinkel\&.\&.\&. "
.ti -1c
.RI "double \fBDotProduct\fP (const \fBmjl::Point\fP &, const \fBmjl::Point\fP &) const"
.br
.ti -1c
.RI "void \fBOut\fP () const"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBmjl::Point\fP \fBorg_\fP"
.br
.ti -1c
.RI "\fBmjl::Point\fP \fBdest_\fP"
.br
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "INTERSECTION mjl::Edge::Cross (const \fBEdge\fP & e, double & t) const\fR [inline]\fP"

.PP
\fBReturns\fP
.RS 4
SKEW_CROSS if the line segments intersect (in constrast with the infinite lines) COLLINEAR, PARALLEL SKEW_NO_CROSS as appropriate t = x value of intersection if intersection occurs 
.RE
.PP

.SS "INTERSECTION mjl::Edge::CrossingPoint (double tolerance, const \fBEdge\fP & f, \fBmjl::Point\fP & p) const"
includes start and end-point into the edge comparison\&. Thus if the intersection is within the tolerance of these nodes, \fBCrossingPoint()\fP will return SKEW_CROSS and it will snap the intersection point to that node\&. 
.SS "INTERSECTION mjl::Edge::Intersect (const \fBEdge\fP & e, double & t) const"

.PP
\fBReturns\fP
.RS 4
SKEW if infinite lines defined by edges crossat a point p COLLINEAR if the lines are the same PARALLEL if the two lines defined by the edges are parallel
.RE
.PP
t is assigned the parametric value (along the line) of where the lines intersect if they do\&. t = x of the intersection 
.SS "INTERSECTION mjl::Edge::SKM_CrossingPoint (double tolerance, const \fBEdge\fP & f, \fBmjl::Point\fP & p) const"
as above but no snapping to points, the intersection point is returned into p\&. 
.SS "EDGE_CLASSIFICATION mjl::Edge::Type (const \fBmjl::Point\fP & p) const"
classifies an edge relative to its location compared with the point p\&. Options are CROSSING, TOUCHING and INESSENTIAL 
.SS "double mjl::Edge::Y (double x) const\fR [inline]\fP"
for a given x value \fBY()\fP returns the corresponding y value on the line defined by the \fBEdge\fP 

.SH "Author"
.PP 
Generated automatically by Doxygen for Open-CSMP++ from the source code\&.


source/modules/discretisation/MJL_Edge.cpp
#include "MJL_Edge.h"
#include <iostream>

using namespace std;

namespace mjl {

void  Edge::Out() const
 {
    cout <<"\norigin: "<< endl;
    org_.Out();
    cout <<"\ndestination: "<< endl;
    dest_.Out();
 }



/**

 classifies an edge relative to its location compared with the
point p. Options are CROSSING, TOUCHING and INESSENTIAL*/
EDGE_CLASSIFICATION  Edge::Type( const mjl::Point& p ) const
 {
    switch( Classify(p) )
      {
         case LEFT:
              return((org_[1]<p[1]) && (p[1]<=dest_[0])) ? CROSSING : INESSENTIAL;
         case RIGHT:
              return((dest_[1]<p[1]) && (p[1]<=org_[1])) ? CROSSING : INESSENTIAL;
         case BETWEEN:
         case ORIGIN:
         case DESTINATION:
              return TOUCHING;
         default:
              return INESSENTIAL;
      }
 }



/// pivots an edge 90o clockwise around it's midpoint
Edge&  Edge::Rot()
 {
    mjl::Point m( org_ + dest_ );
    m *= 0.5;
    mjl::Point v( dest_ - org_ );
    mjl::Point n( v[1], -v[0] ); 
    org_  = m - 0.5 * n;
    dest_ = m + 0.5 * n;
    return *this;
 }




double Edge::DistanceFrom( const mjl::Point& p ) const
 {
    Edge ab = *this;
    ab.Flip().Rot();                 // rotate ab 90o counterclockwise  
    mjl::Point n( ab.dest_ - ab.org_ ); // vector normal to this edge
    n = (1. / n.Length()) * n;  // normalize n
    Edge f( p, p + n );
    double t(std::numeric_limits<double>::quiet_NaN());
    f.Intersect( *this, t ); 
    return t;
 }


bool Edge::IfCloserThanMerge( double tolerance, mjl::Point& p )
 {
    Edge ab = *this;
    ab.Flip().Rot();                 // rotate ab 90o counterclockwise  
    mjl::Point n( ab.dest_ - ab.org_ ); // vector normal to this edge
    n = (1. / n.Length()) * n;  // normalize n
    Edge f( p, p + n );
    double t(std::numeric_limits<double>::quiet_NaN());
    f.Intersect( *this, t ); 
    
    if ( std::fabs(t) <= tolerance ) {
         p = f.Point(t);
         return true;
      }
      
    return false;
 }



/**

includes start and end-point into the edge comparison. Thus if
the intersection is within the tolerance of these nodes, CrossingPoint()
will return SKEW_CROSS and it will snap the intersection point to that node.
*/
INTERSECTION  Edge::CrossingPoint( double tolerance, const Edge& f, mjl::Point& p ) const
 {
    double s, t(std::numeric_limits<double>::quiet_NaN());
    INTERSECTION classe = Intersect( f, s );
    if ((classe == COLLINEAR) || (classe == PARALLEL)) return classe;
    double lene = (dest_ - org_).Length();
    if ((s < -tolerance*lene) || (s > 1.0+tolerance*lene)) return SKEW_NO_CROSS;
    f.Intersect( *this, t );
    double lenf = (f.org_ - f.dest_).Length();
    if ((-tolerance*lenf <= t) && (t <=1.0+tolerance*lenf))
      {
         if      ( t <= tolerance*lenf )      p = f.org_;
         else if ( t >= 1.0*-tolerance*lenf ) p = f.dest_;
         else if ( s <= tolerance*lene )      p = org_;
         else if ( s >= 1.0*-tolerance*lene ) p = dest_;
         else p = f.Point(t);
         return SKEW_CROSS;
      }
    else return SKEW_NO_CROSS;
 }


/**
as above but no snapping to points, the intersection point is returned into p.
*/
INTERSECTION  Edge::SKM_CrossingPoint( double tolerance, const Edge& f, mjl::Point& p ) const
 {
    double s, t(std::numeric_limits<double>::quiet_NaN());
    INTERSECTION classe = Intersect( f, s );
    if ((classe == COLLINEAR) || (classe == PARALLEL)) return classe;
    double lene = (dest_ - org_).Length();
    if ((s < -tolerance*lene) || (s > 1.0+tolerance*lene)) return SKEW_NO_CROSS;
    f.Intersect( *this, t );
    double lenf = (f.org_ - f.dest_).Length();
    if ((-tolerance*lenf <= t) && (t <=1.0+tolerance*lenf))
      {
         p = f.Point(t);
         return SKEW_CROSS;
      }
    else return SKEW_NO_CROSS;
 }



/// see "Mathematische Formeln" p. 170, Schnittwinkel...
double Edge::AngleTo( const Edge& v ) const 
 {
    mjl::Point  a = dest_   - org_;
    mjl::Point  b = v.dest_ - v.org_;
    
    // a b
    // ---
    double ab = a[0]*b[0] + a[1]*b[1];
    
    // |a| . |b|
    // ---------
    double a_dot_b   = std::sqrt( (a[0]*a[0]+a[1]*a[1])*(b[0]*b[0]+b[1]*b[1]) );
    double cos_angle = ab/a_dot_b;

    // if zero intercept
    if ( cos_angle == 0. ) return 90.;
    // if outside of range of 'acos' function
    if ( cos_angle >  1. ) return   0.;
    if ( cos_angle < -1. ) return 180.;
        
    return(180./3.14159265358979323) * std::acos(cos_angle);
    
 } // end AngleTo



/**

@return SKEW if infinite lines defined by edges crossat a point p
COLLINEAR if the lines are the same
PARALLEL if the two lines defined by the edges  are parallel

t is assigned the parametric value (along the line)
of where the lines intersect if they do. t = x of
the intersection
*/
INTERSECTION  Edge::Intersect( const Edge& e, double& t ) const
 {
    mjl::Point n     = mjl::Point((e.dest_-e.org_)[1], (e.org_-e.dest_)[0]);
    double  denom = DotProduct( n, dest_ - org_ );
    if ( denom == 0. )
      {
         LOCATION aclass = org_.Classify( e.org_, e.dest_ );
         if ((aclass == LEFT) || (aclass == RIGHT)) return PARALLEL;
         else return COLLINEAR;
      }
    double num = DotProduct( n, org_ - e.org_ );
    t = -num / denom; // x-value of line intersection
    return SKEW;
 }



/** examines the orientation of 2 edges relative to one-another.
 */
bool  aimsAt( const mjl::Edge& a, const mjl::Edge& b, 
              LOCATION aClass, mjl::INTERSECTION crossType )
  {
     mjl::Point va = a.Destination() - a.Origin();
     mjl::Point vb = b.Destination() - b.Origin();
     
     if ( crossType != COLLINEAR )
       {
          if ( (va[0] * vb[1]) >= (vb[0] * va[1]) ) return( aClass != RIGHT );
          else  return( aClass != LEFT );
       }
     else return( aClass != BEYOND );
 }





/**

Returns the signed angle (0-180o) between an Edge and the
Edge from the beginning point of the Edge and the supplied
point.  
*/
double  signedAngle( const mjl::Point& a, const mjl::Edge& e )
 {
    mjl::Point v = e.Origin()      - a;
    mjl::Point w = e.Destination() - a;
    double  va = v.PolarAngle();
    double  wa = w.PolarAngle();
    if ( (va == -1.) || (wa == -1.) )  return 180.;
    double x = wa - va;
    if ( (x == 180.) || (x == -180.) ) return 180.;
    else if ( x < -180. ) return (x + 360.);
    else if ( x >  180. ) return (x - 360.);
    else return x;
    
 } // end signedAngle






/** Laszlo, p. 162
*/
INTERSECTION  crossingPoint( double tolerance, 
                             const mjl::Edge& e, const mjl::Edge& f, mjl::Point& p )
 {
    double s, t(std::numeric_limits<double>::quiet_NaN());
    
    INTERSECTION classe = e.Intersect( f, s );
    if ((classe == COLLINEAR) || (classe == PARALLEL)) return classe;
    
    double lene = (e.Destination() - e.Origin()).Length();
    
    if ((s < -tolerance*lene) || (s > 1.0+tolerance*lene)) return SKEW_NO_CROSS;
    
    f.Intersect( e, t );
    
    double lenf = (f.Origin() - f.Destination()).Length();
    
    if ((-tolerance*lenf <= t) && (t <=1.0+tolerance*lenf))
      {
         if      ( t <= tolerance*lenf )       p = f.Origin();
         else if ( t >= 1.0 - tolerance*lenf ) p = f.Destination(); 
         else if ( s <= tolerance*lene )       p = e.Origin();
         else if ( s >= 1.0 - tolerance*lene ) p = e.Destination();
         else                                  p = f.Point(t);
         return SKEW_CROSS;
      }
    else return SKEW_NO_CROSS;
 }



EDGE_CLASSIFICATION edgeType( const mjl::Point& a, const Edge& e ) // MJL, p. 119
{
   mjl::Point v = e.Origin();
   mjl::Point w = e.Destination();
   
   switch ( a.Classify(e) ) {
        case LEFT: 
          return ( (v.Y()<a.Y()) and (a.Y()<=w.X()) ) ? CROSSING : INESSENTIAL;
        case RIGHT:
          return ( (w.Y()<a.Y()) and (a.Y()<=v.Y()) ) ? CROSSING : INESSENTIAL;
        case BETWEEN:
        case ORIGIN:
        case DESTINATION:
          return TOUCHING;
        default:
          return INESSENTIAL;
     }
} 


} // end namespace mjl
